### ISSUE TITLE
Record `with` types 

### ISSUE THREAD

**Issue by amarrella** [dhall-lang/dhall-lang#828]
I'm not sure if this came up before, but would it be possible to add a relaxed type checking / projection for records?

e.g. if I have a function of type
```dhall
\(record: { foo : Natural }) -> record
```
and i pass a record of type
```dhall
{ foo = 1, bar = 2 } 
```

currently the typechecker correctly says
```dhall
Error: Wrong type of function argument

{ + bar : …
, …
}
```

Would it be possible to add a `with` annotation to a type, such as 
```dhall
\(record: with { foo : Natural }) -> record
```

which will make the typechecker ignore and remove from the projection the stuff that's not projected? This would allow writing simpler semi-generic functions without resorting to union types, where I require record types `with`  specific fields, and i ignore the others.

One example of where this might be useful:

```dhall
let k8s = https://raw.githubusercontent.com/dhall-lang/dhall-kubernetes/master/types.dhall

let getName = 
   \(object : with { metadata : k8s.ObjectMeta }) -> 
      object.metadata.name
```

I'm not sure if this is a desired feature or not, it's not super critical for me and I know it smells a bit like inheritance, but it feels very ergonomic. 

Let me know what you think :)

**Comment by SiriusStarr** [comment]
The is in essence [row polymorphism](https://en.wikipedia.org/wiki/Row_polymorphism), right?  I personally am a big fan.

**Comment by amarrella** [comment]
@SiriusStarr : TIL row polymorphism, thanks!

What I proposed is a bit weaker than row polymorphism, because we don't preserve the `ρ`, but yeah full row polymorphism is even better! (though i suspect might be harder to implement)

**Comment by SiriusStarr** [comment]
Well, if you're talking about not preserving, then this is sort of like automagic field projection, right?  Which would be neat.

**Comment by sjakobi** [comment]
Thanks for the link @SiriusStarr!

This sounds as if we would be able to "type" record selection and projection then. So we could also have a function like

      \(a : Type) -> \(b : Type) -> \(record : with { fieldA : a, fieldB : b }) -> record.{ fieldA, fieldB }
    : forall(a : Type) -> forall(b : Type) -> (with { fieldA : a, fieldB : b }) -> { fieldA : a, fieldB : b }

To "type" `//` and `/\`, I believe we'd need "more" row polymorphism. (@reactormonk had just mentioned on Slack that he'd like to have this.)

**Comment by sjakobi** [comment]
https://github.com/dhall-lang/dhall-lang/issues/734 and  https://github.com/dhall-lang/dhall-lang/issues/754 look related BTW.

**Comment by sjakobi** [comment]
> #734 and #754 look related BTW.

Also https://github.com/dhall-lang/dhall-lang/issues/116 and https://github.com/dhall-lang/dhall-lang/issues/434.

**Comment by Gabriel439** [comment]
Nix has this in the form of `...` notation for records:

```dhall
\(record: { foo : Natural, ... }) -> record
```

If I remember correctly, the main issue here is that row polymorphism has the extra `ρ` type parameter for a reason.  In a typed language it's not just a nice-to-have: it ensures that unification is well-behaved.  Nix doesn't have this issue because it is untyped.

**Comment by amarrella** [comment]
> If I remember correctly, the main issue here is that row polymorphism has the extra ρ type parameter for a reason. In a typed language it's not just a nice-to-have: it ensures that unification is well-behaved. Nix doesn't have this issue because it is untyped.

@Gabriel439: wouldn't making sure that what's passed is just the projection of the types work? This is different from the nix case. In `nix` if i'm not wrong the type of `record` on the right is effectively the whole record, while here  we could just say
```dhall
\(record : with {foo : Natural}) -> record : { foo :  Natural }
```
which I think is still well typed and safe. 

Having the `ρ` surely is nice for some more advanced stuff such as concatenating operations on the same type though so if that's achievable it's obviously better :)

**Comment by Gabriel439** [comment]
@amarrella: The `ρ` is not just for advanced stuff.  For example: what would be the type of the function that you proposed?

**Comment by amarrella** [comment]
I see what you mean now :) 

How about having a special type which allows this?
Such type would be of kind `* -> *` and accept a record as parameter and the typechecker knows that it should ignore (actually remove) extra fields via projection.

This way, the type of the function above would be 
```
\(record : with {foo : Natural}) -> { foo :  Natural }
```

If that's not possible, I guess preserving the `ρ` is the only way

**Comment by Gabriel439** [comment]
@amarrella: That's somewhat analogous to subtyping, and my understanding the reason why the `ρ` type parameter was introduced for row polymorphism was because unification did not interact well subtyping.

Note that Dhall does not have bidirectional type inference and/or unification, but I expect us to run into similar issues if we don't track things with a type parameter.

I recommend starting from [this post](https://brianmckenna.org/blog/row_polymorphism_isnt_subtyping) as an introduction to the subject.

**Comment by amarrella** [comment]
@Gabriel439: thanks for the link and the explanation! :) I agree that this could be a problem for unification or bidirectional type inference if these were going to be introduced in the language in the future. 

Do you see then `row polymorphism` (as described in the post you linked) as implementable and desirable in dhall?

**Comment by Gabriel439** [comment]
@amarrella: I think if Dhall had bidirectional type inference then row polymorphism would be highly desirable to implement.  In the absence of bidirectional type inference then we wouldn't benefit from row polymorphism, but we could still benefit from "row types" (e.g. a special `Row` kind representing record-only types, such as the type of a valid argument to `//`).

**Comment by trskop** [comment]
I just found myself writing:
```Dhall
λ(t : Type) → λ(v : { foo : Natural } ⩓ t) → v.foo : Natural
```
Which, obviously, can't work.  With `Record Type` it should.

What occurred to me was that if `⩓`, or a different operator, was able to merge noncoliding subrecords, e.g.:
```Dhall
{ foo : Natural } ⩓ { foo : Natural } ≡ { foo : Natural }
{ foo : Natural } ⩓ { foo : Natural, bar : Bool } ≡ { foo : Natural, bar : Bool }
Type error: { foo : Bool } ⩓ { foo : Natural, bar : Bool }
```
We we would get similar properties to row polymorphism, but not row polymorphism.  I'm not versed in type theory enough to understand the implications. Just thought I share the idea.

**Comment by Gabriel439** [comment]
@trskop: You might be interested in the discussion in https://github.com/dhall-lang/dhall-lang/pull/494.  One of the outcomes of that discussion was the introduction of the record projection by type feature, so that your original example would change to:

```dhall
-- ./example.dhall
λ(v : { foo : Natural }) → v.foo : Natural
```

... and to use that function on a larger record you would project out the desired fields, like this:

```dhall
let largerRecord = { foo = 0, bar = True }

let example = ./example.dhall

in  [ example largerRecord.{ foo }
    , example largerRecord.({ foo : Natural })
    ]
```

**Comment by trskop** [comment]
@Gabriel439: Thanks, seems like this was already considered as part of #494.

I probably used an example that wasn't very representative of what would be useful for me.

**Comment by Gabriel439** [comment]
@trskop: Yeah, what you propose would work better if Dhall had type inference.  As mentioned in #494, in the absence of type inference it's less verbose to specify which fields we need to keep rather than which fields we need to ignore

**Comment by mheiber** [comment]
> @amarrella: That's somewhat analogous to subtyping, and my understanding the reason why the `ρ` type parameter was introduced for row polymorphism was because unification did not interact well subtyping.

To support safe `merge` of records, I believe more than ρ is needed: one needs to also track which fields are known *not* to be present. The [MLPolyR thesis](https://arxiv.org/pdf/0910.2654.pdf) uses `{ρ}` to indicate that a row type is missing fields ρ and the [Wikipedia article on row polymorphism](https://en.wikipedia.org/wiki/Row_polymorphism) uses `absent(l)`to indicate the type of a record that does not have a field for `l`.

**Comment by kosta** [comment]
As the last comment here was a more than 2 years ago; and you published the fall-from-grace experiment a little more than a year ago: What is the status of row polymorphism in dhall?

I am currently evaluating dhall for internal use at my company and that is the only feature I'm really missing. (But it can be worked around, I'm asking more out of interest on your thoughts)

**Comment by Gabriella439** [comment]
Yeah, I should probably write something up on what I learned from implementing Grace's type system and what might apply for Dhall

The short answer is that I don't think it will be easy to standardize "good" type inference for Dhall given the present state of the art for implementing type-checkers.  That would include row polymorphism, so I don't think it would be easy to standardize for Dhall.

Slightly longer answer: What I think would need to happen for this to be realistic is for there to be essentially "type-checker generators" available in multiple languages that essentially compile functional logic programming languages to code in each respective host language.  However, that doesn't really exist right now.

**Comment by MonoidMusician** [comment]
I agree. That's essentially why my progress on the problem stalled. I really do want to create that kind of metalanguage, so that we have an executable standard that resembles the logical rules closely _and_ compiles to different backends _and_ enables further optimizations and enhancements on top of the logical presentation. (E.g. I want to keep track of the provenance of each term while typechecking, including accumulating provenance while unifying terms.) But I haven't had time to properly start that work yet, besides coming up with ideas. Plus I have a habit of making all my projects very ambitious and hard to complete, for some reason 😄.
